mmv1/third_party/terraform/functions/name_from_id.go

```diff
@@ -0,0 +1,59 @@
+package functions
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/hashicorp/terraform-plugin-framework/function"
+)
+
+var _ function.Function = NameFromIdFunction{}
```

Comment:
Side question: in your experience does "name" usually refer to the "short name"? I think we've seen both long/short name when referencing "name" internally, so sometimes there is confusion. I don't have a lot of references externally for what folks usually mean by "name".

Reply:
This has been confusing to me too.

Here are things that make me think of "name" as referring to the short name of a resource:

1) A guide in the Google provider's docs says ([link](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/getting_started#linking-gcp-resources:~:text=Like%20this%20VM%20instance%2C%20nearly%20every%20GCP%20resource%20will%20have%20a%20name%20field.%20They%20are%20used%20as%20a%20short%20way%20to%20identify%20resources%2C%20and%20a%20resource%27s%20display%20name%20in%20the%20Cloud%20Console%20will%20be%20the%20one%20defined%20in%20the%20name%20field.)):

> Like this VM instance, nearly every GCP resource will have a `name` field. They are used as a short way to identify resources, and a resource's display name in the Cloud Console will be the one defined in the `name` field.

Equating display name to `name` makes me think it's definitely referring to the short-form name.

2) Our documentation for the provider shows [examples with `name` (or equivalent) arguments being set like `name = "my-instance"`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance#example-usage). If an API returns a longer value after the resource is created there are often DiffSuppress functions like [`CompareSelfLinkOrResourceName`](https://github.com/hashicorp/terraform-provider-google/blob/099222630bbeac3ede04c7bf05217c3be2f32917/google/tpgresource/self_link_helpers.go#L40-L57) that ensures an end user only sees the name field as being the short-form version.

3) Our codebase includes code that refers to the last bit of a string that contains `/` characters as being the name. As well as [`CompareSelfLinkOrResourceName`](https://github.com/hashicorp/terraform-provider-google/blob/099222630bbeac3ede04c7bf05217c3be2f32917/google/tpgresource/self_link_helpers.go#L40-L57) above there are other helpers like [the `GetResourceNameFromSelfLink` helper function](https://github.com/hashicorp/terraform-provider-google/blob/099222630bbeac3ede04c7bf05217c3be2f32917/google/tpgresource/self_link_helpers.go#L85-L88).



In the past my understanding was that:
- self link = `https://cloudkms.googleapis.com/v1/projects/my-project/locations/us-central1/keyRings/my-key-ring/cryptoKeys/my-crypto-key`
    - [provider docs describe this in this guide](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/getting_started#linking-gcp-resources)
- id = `projects/my-project/locations/us-central1/keyRings/my-key-ring/cryptoKeys/my-crypto-key`
    -  `id` attributes for resources are documented to have this format, [e.g. for BigQuery datasets](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_dataset#id)
- name = `my-crypto-key`
     - [provider docs describe this in this guide](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/getting_started#linking-gcp-resources)


_But_ then Riley referred me to this documentation about[ "full resource names" ](https://google.aip.dev/122#full-resource-names), i.e. `//cloudkms.googleapis.com/v1/projects/my-project/locations/us-central1/keyRings/my-key-ring/cryptoKeys/my-crypto-key`, which adds confusion. Compared to the provider documentation etc I don't know how much impact the AIP docs change what end users think about the concept of a name.

Reply:
Sorry for the essay! 😅 If you think that the function name could be named in a more accurate way I'm happy to change it now, before it's released. Perhaps `short_name_from_id`? What do you think?

Reply:
Thanks for compiling that, it definitely looks like we use `name` in the short form more often than not externally. So I think it is fine to leave it as is!

Reply:
Perfect, thanks 😁 If any strong reason to rename the function comes up it'll be relatively easy to create an identical function with a different name and then deprecate the older version in future